includes/html/api_functions.inc.php

```diff
@@ -2193,51 +2195,38 @@ function list_mpls_services(Illuminate\Http\Request $request)
     $sql_params = [];
     $hostname = $request->get('hostname');
     $device_id = ctype_digit($hostname) ? $hostname : getidbyname($hostname);
-    if (is_numeric($device_id)) {
-        $permission = check_device_permission($device_id);
-        if ($permission !== true) {
-            return $permission;
-        }
-        $sql = ' AND `devices`.`device_id`=?';
-        $sql_params = [$device_id];
-    }
     
-    if (! Auth::user()->hasGlobalRead()) {
-        $sql .= ' AND `mpls_services`.`device_id` IN (SELECT device_id FROM devices_perms WHERE user_id = ?)';
-        $sql_params[] = Auth::id();
+    $mpls_services = MplsService::hasAccess(Auth::user())->get();
```

Comment:
Instead of fetching all MplsService rows then discarding  the ones you don't want try this
```php
$mpls_services = MplsService::hasAccess(Auth::user())->when($device_id, function ($query, $device_id) {
    return $query->where('device_id', $device_id);
})->get();
```